src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Globalization;
+
+namespace NuGet.SolutionRestoreManager
+{
+    [Export(typeof(IVsNuGetProjectUpdateEvents))]
+    [Export(typeof(IVsNuGetProgressReporter))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class VsRestoreProgressEvents : IVsNuGetProjectUpdateEvents, IVsNuGetProgressReporter
+    {
+        public event SolutionRestoreEventHandler SolutionRestoreStarted;
+        public event SolutionRestoreEventHandler SolutionRestoreFinished;
+        public event ProjectUpdateEventHandler ProjectUpdateStarted;
+        public event ProjectUpdateEventHandler ProjectUpdateFinished;
+
+        public void EndProjectUpdate(string projectName, IReadOnlyList<string> updatedFiles)
+        {
+            if (projectName == null)
+            {
+                throw new ArgumentNullException(nameof(projectName));
+            }
+
+            if (updatedFiles == null)
+            {
+                throw new ArgumentNullException(nameof(updatedFiles));
+            }
+
+            if (ProjectUpdateFinished != null)
+            {
+                foreach (var handler in ProjectUpdateFinished.GetInvocationList())
+                {
+                    try
+                    {
+                        handler.DynamicInvoke(projectName, updatedFiles);
+                    }
+                    catch { }
+                }
```

Comment:
I see this piece of code repeated several times in this PR. Can we abstract this logic into a utility method?

Reply:
Things are sort/kind of duplicated only once, given the parameters. 
This is code that we'll pretty much never change, so not too worried about the duplication.